🔢 NumPy Tutorial › NumPy ufunc &nbsp;·&nbsp; *Bài 38/44*

[« NumPy Summations](037_ufunc_summations.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../../README.md) &nbsp;•&nbsp; [NumPy Differences »](039_ufunc_differences.ipynb)

# NumPy Products

> 📖 **Học song song trên W3Schools:** [numpy_ufunc_products.asp](https://www.w3schools.com/python/numpy/numpy_ufunc_products.asp)
>
> 🎯 **Trong bài này:** Products · Product Over an Axis · Cumulative Product

## Products — Tích

Dùng hàm `prod()` để tính **tích** các phần tử:

In [1]:
import numpy as np

arr = np.array([1, 2, 3, 4])

x = np.prod(arr)

print(x)

24


Kết quả: `24` vì 1 × 2 × 3 × 4 = 24.

Tích của **hai mảng**:

In [2]:
import numpy as np

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

x = np.prod([arr1, arr2])

print(x)

40320


Kết quả: `40320` vì 1×2×3×4×5×6×7×8 = 40320.

## Product Over an Axis — Tích theo trục

In [3]:
import numpy as np

arr1 = np.array([1, 2, 3, 4])
arr2 = np.array([5, 6, 7, 8])

newarr = np.prod([arr1, arr2], axis=1)

print(newarr)

[  24 1680]


## Cumulative Product — Tích tích luỹ

Tích tích luỹ của `[1, 2, 3, 4]` là `[1, 1*2, 1*2*3, 1*2*3*4]` = `[1, 2, 6, 24]`. Dùng `cumprod()`:

In [4]:
import numpy as np

arr = np.array([5, 6, 7, 8])

newarr = np.cumprod(arr)

print(newarr)

[   5   30  210 1680]


---

## 🧠 Ghi nhớ nhanh

> - `np.prod(arr)` → tích mọi phần tử; `axis=1` → tích từng hàng.
> - `np.cumprod(arr)` → tích **tích luỹ**.

---

## 📝 Exercise — Bài tập nhanh

**❓ Câu hỏi:** `np.prod([2, 3, 4])` bằng bao nhiêu?

- **A.** 234
- **B.** 6
- **C.** 24
- **D.** 9

<details>
<summary>👉 Xem đáp án</summary>

✅ **Đáp án: C** — 24

</details>

---

[« NumPy Summations](037_ufunc_summations.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../../README.md) &nbsp;•&nbsp; [NumPy Differences »](039_ufunc_differences.ipynb)